src/v/archival/ntp_archiver_service.cc

```diff
@@ -508,6 +508,48 @@ ss::future<> ntp_archiver::sync_manifest_until_term_change() {
     }
 }
 
+ss::future<> ntp_archiver::unsafe_reset_metadata(ss::input_stream<char> is) {
+    // Deserialize outside of lock.
+    cloud_storage::partition_manifest req_m{};
+    co_await req_m.update(std::move(is));
+
+    // We're about to blow away the manifest. Lock it so the upload loop et al
+    // don't try uploading while this is happening.
+    auto sync_timeout = config::shard_local_cfg()
+                          .cloud_storage_metadata_sync_timeout_ms.value();
+
+    auto deadline = ss::lowres_clock::now() + sync_timeout;
+    auto builder = _parent.archival_meta_stm()->batch_start(deadline, _as);
+    builder.reset_metadata();
+
+    // Add segments. Note that we only add, and omit the replaced segments.
+    // Presumably this is being done in the context of a last-ditch effort to
+    // restore functionality to a partition. Replaced segments are likely
+    // unimportant.
+    // TODO: think about archival section.
+    std::vector<cloud_storage::segment_meta> segments;
+    segments.reserve(manifest().size());
+    for (const auto& s : req_m) {
+        segments.emplace_back(s);
+    }
+    builder.add_segments(std::move(segments));
```

Comment:
Should we batch calls to `add_segments` in order to prevent huge allocations here?